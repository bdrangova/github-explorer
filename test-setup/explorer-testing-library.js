import React from 'react';
import { IntlProvider } from 'react-intl';
import { render } from '@testing-library/react';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import { createStore } from 'redux';
import { Provider } from 'react-redux';
import rootReducer from 'src/store';
import { messages } from 'src/locales';

import { FirebaseAuth } from 'src/providers/Auth';

const renderWithRouter = (
  ui,
  { route = '/', renderOptions = {} } = {},
) => {
  const history = createMemoryHistory({ initialEntries: [route] });
  const utils = render(renderRouter(ui, history), renderOptions);

  return {
    ...utils,
    history,
  };
};

const renderRouter = (ui, history) => {
  return <Router history={history}>{ui}</Router>;
};

const renderWithRedux = (
  ui,
  { initialState = {}, renderOptions = {} } = {},
) => {
  const utils = render(
    renderRedux(ui, initialState),
    ...renderOptions,
  );
  return {
    ...utils,
  };
};

const renderRedux = (ui, initialState = {}) => {
  const store = createStore(rootReducer, initialState);
  return <Provider store={store}>{ui}</Provider>;
};

const renderWithAuth = (
  ui,
  { value = {}, renderOptions = {} } = {},
) => {
  const utils = render(renderAuth(ui, value));
  return { ...utils };
};

const renderAuth = (ui, value = {}) => {
  return (
    <FirebaseAuth.Provider value={value}>{ui}</FirebaseAuth.Provider>
  );
};

const renderWithIntl = (ui, { renderOptions = {} } = {}) => {
  const utils = render(renderIntl(ui, renderOptions), renderOptions);
  return { ...utils };
};

const renderIntl = (ui) => {
  return (
    <IntlProvider locale="en" messages={messages['en']}>
      {ui}
    </IntlProvider>
  );
};

const renderFullstack = (
  ui,
  initialState = {},
  route = '/',
  value = {},
) => {
  const history = createMemoryHistory({ initialEntries: [route] });
  return renderRedux(
    renderIntl(renderRouter(renderAuth(ui, value), history)),
    initialState,
  );
};

const renderWithFullstack = (
  ui,
  initialState = {},
  route = '/',
  authValue = {},
  renderOptions = {},
) => {
  return render(
    renderFullstack(ui, initialState, route, authValue),
    renderOptions,
  );
};

const renderWithIntlAndRouter = (
  ui,
  route = '/',
  renderOptions = {},
) => {
  const history = createMemoryHistory({ initialEntries: [route] });
  return renderWithIntl(renderRouter(ui, history), renderOptions);
};

export {
  renderRouter,
  renderRedux,
  renderAuth,
  renderIntl,
  renderWithRouter,
  renderWithRedux,
  renderWithAuth,
  renderWithIntl,
  renderWithIntlAndRouter,
  renderWithFullstack,
};
