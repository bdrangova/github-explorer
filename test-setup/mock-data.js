const RepositoryMock = {
  id: 1234,
  name: 'react',
  full_name: 'facebook/react',
  description: 'Library for building UIs',
  stargazers_count: 150000,
  language: 'JavaScript',
  updated_at: '',
  owner: {
    login: 'facebook',
  },
  private: false,
};

const RepositoryArrayMock = [
  {
    id: 1234,
    name: 'react',
    full_name: 'facebook/react',
    description: 'Library for building UIs',
    stargazers_count: 150000,
    language: 'JavaScript',
    updated_at: '',
    owner: {
      login: 'facebook',
    },
    private: false,
  },
  {
    id: 46456,
    name: 'vue',
    full_name: 'vuejs/vue',
    description: 'Progressive framework',
    stargazers_count: 16000,
    language: 'JavaScript',
    updated_at: '',
    owner: {
      login: 'vuejs',
    },
    private: false,
  },
];

const MockedGithubSearchResponse = {
  total_count: 2,
  incomplete_results: false,
  items: [
    {
      id: 11730342,
      name: 'vue',
      full_name: 'vuejs/vue',
      private: false,
      owner: {
        login: 'vuejs',
        id: 6128107,
        avatar_url:
          'https://avatars1.githubusercontent.com/u/6128107?v=4',
      },
      html_url: 'https://github.com/vuejs/vue',
      description:
        'üññ Vue.js is a progressive, incrementally-adoptable JavaScript framework for building UI on the web.',
      updated_at: '2020-06-02T08:29:16Z',

      stargazers_count: 165128,
      watchers_count: 165128,
      language: 'JavaScript',
      forks_count: 25032,
      open_issues_count: 509,
      forks: 25032,
    },
    {
      id: 77189043,
      name: 'vue2-elm',
      full_name: 'bailicangdu/vue2-elm',
      private: false,
      owner: {
        login: 'bailicangdu',
        id: 20297227,
        avatar_url:
          'https://avatars2.githubusercontent.com/u/20297227?v=4',
      },
      html_url: 'https://github.com/bailicangdu/vue2-elm',
      description:
        'Âü∫‰∫é vue2 + vuex ÊûÑÂª∫‰∏Ä‰∏™ÂÖ∑Êúâ 45 ‰∏™È°µÈù¢ÁöÑÂ§ßÂûãÂçïÈ°µÈù¢Â∫îÁî®',
      updated_at: '2020-06-02T08:13:35Z',
      size: 29812,
      stargazers_count: 33331,
      watchers_count: 33331,
      language: 'Vue',
      forks_count: 11052,
      open_issues_count: 72,
    },
  ],
};

export {
  RepositoryMock,
  RepositoryArrayMock,
  MockedGithubSearchResponse,
};
