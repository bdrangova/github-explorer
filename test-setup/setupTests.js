import '@testing-library/jest-dom/extend-expect';
require('jest-fetch-mock').enableMocks();
require('../__mocks__/browserMocks');
