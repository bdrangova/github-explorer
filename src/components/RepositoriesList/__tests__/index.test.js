import React from 'react';
import { renderWithIntlAndRouter } from 'test-setup/explorer-testing-library';
import { fireEvent } from '@testing-library/react';
import { RepositoryArrayMock } from 'test-setup/mock-data';
import RepositoryList from '../index';

const pagerMock = {
  current: 1,
};

const setLayoutMock = jest.fn();
const handlePageChangeMock = jest.fn();

const RepoListMockData = {
  repos: RepositoryArrayMock,
  grid: true,
  pager: pagerMock,
  loading: false,
  duration: 0.45,
  setLayout: setLayoutMock,
  handlePageChange: handlePageChangeMock,
};

describe('Repositories List', () => {
  it('repositories list snapshot', () => {
    const { container } = renderWithIntlAndRouter(
      <RepositoryList {...RepoListMockData} />,
    );
    expect(container).toMatchSnapshot();
  });

  it('change layout type from grid to list', async () => {
    const { getByTestId } = renderWithIntlAndRouter(
      <RepositoryList {...RepoListMockData} />,
    );
    const setListBtn = getByTestId('set-list-btn');
    fireEvent.click(setListBtn);

    expect(setLayoutMock).toHaveBeenCalledTimes(1);
    expect(setLayoutMock).toHaveBeenCalledWith('list');
  });
});
