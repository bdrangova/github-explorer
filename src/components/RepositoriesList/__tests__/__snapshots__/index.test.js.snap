// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Repositories List repositories list snapshot 1`] = `
<div>
  <div
    class="MuiContainer-root container MuiContainer-maxWidthLg"
  >
    <div
      class="MuiBox-root MuiBox-root-1 box"
    >
      <div
        class="MuiBox-root MuiBox-root-2 box"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-colorPrimary MuiSvgIcon-fontSizeSmall"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M6 2v6h.01L6 8.01 10 12l-4 4 .01.01H6V22h12v-5.99h-.01L18 16l-4-4 4-3.99-.01-.01H18V2H6z"
          />
        </svg>
        <span
          class="MuiTypography-root MuiTypography-caption"
        >
          0.45
           
          seconds
        </span>
      </div>
      <div
        aria-label="contained primary button group"
        class="MuiButtonGroup-root MuiButtonGroup-contained"
        role="group"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedContained MuiButtonGroup-groupedContainedHorizontal MuiButtonGroup-groupedContained Mui-disabled Mui-disabled"
          data-testid="set-grid-btn"
          disabled=""
          tabindex="-1"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M4 8h4V4H4v4zm6 12h4v-4h-4v4zm-6 0h4v-4H4v4zm0-6h4v-4H4v4zm6 0h4v-4h-4v4zm6-10v4h4V4h-4zm-6 4h4V4h-4v4zm6 6h4v-4h-4v4zm0 6h4v-4h-4v4z"
              />
            </svg>
          </span>
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedContained MuiButtonGroup-groupedContainedHorizontal MuiButtonGroup-groupedContained"
          data-testid="set-list-btn"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M3 13h2v-2H3v2zm0 4h2v-2H3v2zm0-8h2V7H3v2zm4 4h14v-2H7v2zm0 4h14v-2H7v2zM7 7v2h14V7H7z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <div
      class="MuiGrid-root grid MuiGrid-container MuiGrid-spacing-xs-3 MuiGrid-justify-xs-space-between"
      data-testid="results-display"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-4 MuiGrid-grid-md-4"
      >
        <div
          class="MuiPaper-root MuiCard-root card MuiPaper-elevation1 MuiPaper-rounded"
        >
          <button
            class="MuiButtonBase-root MuiCardActionArea-root card"
            tabindex="0"
            type="button"
          >
            <a
              href="/repos/facebook/react"
              style="text-decoration: none;"
            >
              <div
                class="MuiCardContent-root"
              >
                <div
                  class="stats"
                >
                  <h6
                    class="MuiTypography-root owner MuiTypography-subtitle1 MuiTypography-colorTextSecondary"
                    data-testid="owner-name"
                  >
                    facebook
                  </h6>
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-colorPrimary"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm-1 17.93c-3.95-.49-7-3.85-7-7.93 0-.62.08-1.21.21-1.79L9 15v1c0 1.1.9 2 2 2v1.93zm6.9-2.54c-.26-.81-1-1.39-1.9-1.39h-1v-3c0-.55-.45-1-1-1H8v-2h2c.55 0 1-.45 1-1V7h2c1.1 0 2-.9 2-2v-.41c2.93 1.19 5 4.06 5 7.41 0 2.08-.8 3.97-2.1 5.39z"
                    />
                  </svg>
                </div>
                <h2
                  class="MuiTypography-root title MuiTypography-h6 MuiTypography-noWrap"
                  data-testid="repo-name"
                >
                  react
                </h2>
                <p
                  class="MuiTypography-root MuiTypography-body2 MuiTypography-noWrap MuiTypography-gutterBottom"
                >
                  Library for building UIs
                </p>
                <div
                  class="stats"
                >
                  <h6
                    class="MuiTypography-root typography MuiTypography-subtitle1 MuiTypography-colorTextSecondary"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-colorPrimary"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                      />
                    </svg>
                    150K
                  </h6>
                  <h6
                    class="MuiTypography-root typography MuiTypography-subtitle1 MuiTypography-colorTextSecondary"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-colorPrimary"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M9.4 16.6L4.8 12l4.6-4.6L8 6l-6 6 6 6 1.4-1.4zm5.2 0l4.6-4.6-4.6-4.6L16 6l6 6-6 6-1.4-1.4z"
                      />
                    </svg>
                    JavaScript
                  </h6>
                </div>
                <h6
                  class="MuiTypography-root pos MuiTypography-subtitle1 MuiTypography-colorTextSecondary"
                >
                  Updated on Invalid Date at Invalid Date
                </h6>
              </div>
            </a>
            <span
              class="MuiCardActionArea-focusHighlight"
            />
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-4 MuiGrid-grid-md-4"
      >
        <div
          class="MuiPaper-root MuiCard-root card MuiPaper-elevation1 MuiPaper-rounded"
        >
          <button
            class="MuiButtonBase-root MuiCardActionArea-root card"
            tabindex="0"
            type="button"
          >
            <a
              href="/repos/vuejs/vue"
              style="text-decoration: none;"
            >
              <div
                class="MuiCardContent-root"
              >
                <div
                  class="stats"
                >
                  <h6
                    class="MuiTypography-root owner MuiTypography-subtitle1 MuiTypography-colorTextSecondary"
                    data-testid="owner-name"
                  >
                    vuejs
                  </h6>
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-colorPrimary"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm-1 17.93c-3.95-.49-7-3.85-7-7.93 0-.62.08-1.21.21-1.79L9 15v1c0 1.1.9 2 2 2v1.93zm6.9-2.54c-.26-.81-1-1.39-1.9-1.39h-1v-3c0-.55-.45-1-1-1H8v-2h2c.55 0 1-.45 1-1V7h2c1.1 0 2-.9 2-2v-.41c2.93 1.19 5 4.06 5 7.41 0 2.08-.8 3.97-2.1 5.39z"
                    />
                  </svg>
                </div>
                <h2
                  class="MuiTypography-root title MuiTypography-h6 MuiTypography-noWrap"
                  data-testid="repo-name"
                >
                  vue
                </h2>
                <p
                  class="MuiTypography-root MuiTypography-body2 MuiTypography-noWrap MuiTypography-gutterBottom"
                >
                  Progressive framework
                </p>
                <div
                  class="stats"
                >
                  <h6
                    class="MuiTypography-root typography MuiTypography-subtitle1 MuiTypography-colorTextSecondary"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-colorPrimary"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                      />
                    </svg>
                    16K
                  </h6>
                  <h6
                    class="MuiTypography-root typography MuiTypography-subtitle1 MuiTypography-colorTextSecondary"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-colorPrimary"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M9.4 16.6L4.8 12l4.6-4.6L8 6l-6 6 6 6 1.4-1.4zm5.2 0l4.6-4.6-4.6-4.6L16 6l6 6-6 6-1.4-1.4z"
                      />
                    </svg>
                    JavaScript
                  </h6>
                </div>
                <h6
                  class="MuiTypography-root pos MuiTypography-subtitle1 MuiTypography-colorTextSecondary"
                >
                  Updated on Invalid Date at Invalid Date
                </h6>
              </div>
            </a>
            <span
              class="MuiCardActionArea-focusHighlight"
            />
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
    <div
      class="MuiBox-root MuiBox-root-3 container"
    >
      <nav
        aria-label="pagination navigation"
        class="MuiPagination-root"
      >
        <ul
          class="MuiPagination-ul"
        >
          <li>
            <button
              aria-label="Go to previous page"
              class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page MuiPaginationItem-textPrimary Mui-disabled Mui-disabled"
              disabled=""
              tabindex="-1"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiPaginationItem-icon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
                />
              </svg>
            </button>
          </li>
          <li>
            <button
              aria-current="true"
              aria-label="page 1"
              class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page MuiPaginationItem-textPrimary Mui-selected"
              tabindex="0"
              type="button"
            >
              1
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </li>
          <li>
            <button
              aria-label="Go to next page"
              class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page MuiPaginationItem-textPrimary Mui-disabled Mui-disabled"
              disabled=""
              tabindex="-1"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiPaginationItem-icon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
                />
              </svg>
            </button>
          </li>
        </ul>
      </nav>
    </div>
  </div>
</div>
`;
